src/northbridge/amd/amdfam10/northbridge.c

```diff
@@ -60,14 +675,572 @@ static struct device_operations pci_domain_ops = {
+	dev_mc = dev_mc ? dev_mc : pcidev_on_root(0x18, 0);
+
+	if (dev_mc && dev_mc->bus) {
+		printk(BIOS_DEBUG, "%s found\n", dev_path(dev_mc));
+		pci_domain = dev_mc->bus->dev;
+		if (pci_domain && (pci_domain->path.type == DEVICE_PATH_DOMAIN)) {
+			if ((pci_domain->link_list) &&
+			    (pci_domain->link_list->children == dev_mc)) {
+				printk(BIOS_DEBUG, "%s move to ",dev_path(dev_mc));
+				dev_mc->bus->secondary = CONFIG_CBB; // move to 0xff
+				printk(BIOS_DEBUG, "%s\n",dev_path(dev_mc));
+				while (dev_mc) {
+					printk(BIOS_DEBUG, "%s move to ",dev_path(dev_mc));
+					dev_mc->path.pci.devfn -= PCI_DEVFN(0x18,0);
+					printk(BIOS_DEBUG, "%s\n",dev_path(dev_mc));
+					dev_mc = dev_mc->sibling;
+				}
+			}
+		}
+	}
+}
+
+static void remap_bsp_lapic(struct bus *cpu_bus)
+{
+	struct device_path cpu_path;
+	struct device *cpu;
+	u32 bsp_lapic_id = lapicid();
+
+	if (bsp_lapic_id) {
+		cpu_path.type = DEVICE_PATH_APIC;
+		cpu_path.apic.apic_id = 0;
+		cpu = find_dev_path(cpu_bus, &cpu_path);
+		if (cpu)
+			cpu->path.apic.apic_id = bsp_lapic_id;
+	}
+}
+
+static unsigned int get_num_siblings(void)
+{
+	unsigned int siblings;
+	unsigned int ApicIdCoreIdSize = (cpuid_ecx(0x80000008)>>12 & 0xf);
+
+	if (ApicIdCoreIdSize) {
+		siblings = (1 << ApicIdCoreIdSize) - 1;
+	} else {
+		siblings = 3; //quad core
+	}
+
+	return siblings;
+}
+
+static void setup_cdb_links(unsigned int busn, unsigned int devn, struct bus *pbus)
+{
+	struct device *cdb_dev;
+	/* Find the cpu's pci device */
+	cdb_dev = dev_find_slot(busn, PCI_DEVFN(devn, 0));
+	if (!cdb_dev) {
+		/* If I am probing things in a weird order
+			* ensure all of the cpu's pci devices are found.
+			*/
+		int fn;
+		for (fn = 0; fn <= 5; fn++) { //FBDIMM?
+			cdb_dev = pci_probe_dev(NULL, pbus,
+				PCI_DEVFN(devn, fn));
+		}
+	}
+
+
+	/* Ok, We need to set the links for that device.
+		* otherwise the device under it will not be scanned
+		*/
+	cdb_dev = dev_find_slot(busn, PCI_DEVFN(devn, 0));
+	if (cdb_dev)
+		add_more_links(cdb_dev, 4);
+
+	cdb_dev = dev_find_slot(busn, PCI_DEVFN(devn, 4));
+	if (cdb_dev)
+		add_more_links(cdb_dev, 4);
+}
+
+static int get_num_cores(unsigned int busn, unsigned int devn, int *cores_found,
+			 int *enable_node)
+{
+	int j;
+	struct device *cdb_dev;
+	unsigned int siblings = get_num_siblings();
+
+	*cores_found = 0; // one core
+	if (is_fam15h())
+		cdb_dev = dev_find_slot(busn, PCI_DEVFN(devn, 5));
+	else
+		cdb_dev = dev_find_slot(busn, PCI_DEVFN(devn, 3));
+
+	*enable_node = cdb_dev && cdb_dev->enabled;
+	if (*enable_node) {
+		if (is_fam15h()) {
+			*cores_found = pci_read_config32(cdb_dev, 0x84) & 0xff;
+		} else {
+			j = pci_read_config32(cdb_dev, 0xe8);
+			*cores_found = (j >> 12) & 3; // dev is func 3
+			if (siblings > 3)
+				*cores_found |= (j >> 13) & 4;
+		}
+		printk(BIOS_DEBUG, "  %s siblings=%d\n", dev_path(cdb_dev), *cores_found);
+	}
+
+	if (siblings > *cores_found)
+		siblings = *cores_found;
```

Comment:
Local variable is changed, but this isn't propagated anywhere. Most notably, `get_apic_id()` doesn't see this.